### PR TITLE
os-util: several cleanups and introduce _at() variants of os-release parsers

### PR DESCRIPTION
**Pull request by yuwata**
preparation for #26848.

### CODE REVIEW

File: src/basic/os-util.c [hunk 1]
```diff
@@ -114,205 +122,215 @@ static int extension_release_strict_xattr_value(int extension_release_fd, const
...
 
-        return 0;
+int open_os_release(const char *root, char **ret_path, int *ret_fd) {
+        _cleanup_close_ int rfd = -EBADF;
+
+        rfd = open(empty_to_root(root), O_CLOEXEC | O_DIRECTORY | O_PATH);
+        if (rfd < 0)
+                return -errno;
+
+        return open_os_release_at(rfd, ret_path, ret_fd);
```
Comment (daandemeyer) [inline src/basic/os-util.c]: We need to prefix with root here no?

File: src/basic/env-file.c [hunk 2]
```diff
@@ -359,6 +359,27 @@ int parse_env_filev(
         return r;
 }
 
+int parse_env_file_fdv(
+                int fd,
+                const char *fname,
+                va_list ap) {
```
Comment (daandemeyer) [inline src/basic/env-file.c]: This can be one line no?

File: src/basic/os-util.c [hunk 3]
```diff
@@ -156,6 +156,9 @@ int open_extension_release(
         if (!extension)
                 return open_os_release(root, ret_path, ret_fd);
 
+        if (!IN_SET(image_class, IMAGE_SYSEXT, IMAGE_CONFEXT))
+                return -EINVAL;
```
Comment (daandemeyer) [inline src/basic/os-util.c:191]: Shouldn't this be an assertion?

File: src/basic/os-util.c [hunk 4]
```diff
@@ -242,24 +255,78 @@ int open_extension_release(
...
+                char **ret_path,
+                int *ret_fd) {
+
+        _cleanup_close_ int rfd = -EBADF;
+
+        rfd = open(empty_to_root(root), O_CLOEXEC | O_DIRECTORY | O_PATH);
+        if (rfd < 0)
+                return -errno;
+
+        return open_extension_release_at(rfd, image_class, extension, relax_extension_release_check, ret_path, ret_fd);
```
Comment (daandemeyer) [inline src/basic/os-util.c]: Ditto, we have to prefix with root here no?

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Looks good, but a few comments

File: src/basic/os-util.c (hunk 1 above)
Comment (yuwata) [reply src/basic/os-util.c]: What do you mean?

Comment (daandemeyer) [reply src/basic/os-util.c]: ret_path will be relative to root here, but we need it to be returned as an absolute path no?

Comment (yuwata) [reply src/basic/os-util.c]: Ah, indeed.

**Comment by yuwata** [comment]
@DaanDeMeyer Thank you for the review. All requests are addressed. PTAL.

File: src/basic/os-util.c (hunk 3 above)
Comment (yuwata) [reply src/basic/os-util.c:191]: Making this assertion causes ABRT in machined.

**Comment by yuwata** [comment]
@DaanDeMeyer The previous CI failure seems to cause the suggestion https://github.com/systemd/systemd/pull/27186#discussion_r1161600736. I reverted the suggestion and use the original version. PTAL.

**Review by DaanDeMeyer — APPROVED** [review]

**Pull request merged**